File: example/templates/addons/desktop_widgets/mediawidget/src/ui/components/marquee.ts

import Gtk from 'gi://Gtk?version=3.0';
import GLib from 'gi://GLib?version=2.0';

export type MarqueeDirection = 'left' | 'right';

export class MarqueeComponent {
  public container: Gtk.ScrolledWindow;
  private label1: Gtk.Label;
  private label2: Gtk.Label;
  private adjustment: Gtk.Adjustment;
  private scrollBox: Gtk.Box;

  private offset = 0;
  private waiting = 0;
  private enabled = false;
  private scrollLimit = 0;
  private direction: MarqueeDirection;
  private containerWidth = 0;

  // Config constants
  private readonly PAUSE_FRAMES = 100;
  private readonly GAP = 60;
  private readonly SPEED = 1;

  constructor(className: string, direction: MarqueeDirection = 'left') {
    this.direction = direction;

    this.label1 = new Gtk.Label({
      label: '...',
      halign: Gtk.Align.START,
      wrap: false,
      lines: 1,
    });
    this.label1.get_style_context().add_class(className);

    this.label2 = new Gtk.Label({
      label: '...',
      halign: Gtk.Align.START,
      wrap: false,
      lines: 1,
    });
    this.label2.get_style_context().add_class(className);

    this.scrollBox = new Gtk.Box({
      orientation: Gtk.Orientation.HORIZONTAL,
      spacing: this.GAP,
    });
    this.scrollBox.pack_start(this.label1, false, false, 0);
    this.scrollBox.pack_start(this.label2, false, false, 0);
    this.scrollBox.halign = Gtk.Align.START; // Always left align

    this.adjustment = new Gtk.Adjustment({
      lower: 0,
      upper: 10000,
      step_increment: 1,
      page_size: 100,
    });

    this.container = new Gtk.ScrolledWindow({
      hscrollbar_policy: Gtk.PolicyType.EXTERNAL,
      vscrollbar_policy: Gtk.PolicyType.NEVER,
      hadjustment: this.adjustment,
    });

    // Critical: Disable natural width propagation to prevent expansion
    this.container.set_propagate_natural_width(false);
    this.container.set_propagate_natural_height(false);

    if (className === 'title') {
      this.container.get_style_context().add_class('title-scroll');
    }

    this.container.add(this.scrollBox);

    // Capture container width once allocated
    this.container.connect('size-allocate', (widget, alloc) => {
      if (this.containerWidth === 0) {
        this.containerWidth = alloc.width;
      }
    });
  }

  public setText(text: string) {
    this.label1.label = text;
    this.label2.label = text;
    this.reset();
  }

  private reset() {
    this.offset = 0;
    this.waiting = this.PAUSE_FRAMES;
    this.adjustment.set_value(0);
    this.enabled = false;
    this.label2.hide();
    this.scrollBox.halign = Gtk.Align.START; // Always left align

    // Check if we need to scroll after layout settles
    GLib.timeout_add(GLib.PRIORITY_DEFAULT, 100, () => {
      const scrollAlloc = this.container.get_allocation();
      const [, nat] = this.label1.get_preferred_width();
      const labelWidth = nat;

      if (labelWidth > scrollAlloc.width) {
        this.enabled = true;
        this.scrollLimit = labelWidth + this.GAP;
        this.label2.show();

        // For right direction, start at max offset
        if (this.direction === 'right') {
          this.offset = this.scrollLimit;
          this.adjustment.set_value(this.offset);
        }
      }
      return false;
    });
  }

  public tick() {
    if (!this.enabled) return;

    if (this.waiting > 0) {
      this.waiting--;
      return;
    }

    if (this.direction === 'left') {
      this.offset += this.SPEED;
      if (this.offset >= this.scrollLimit) {
        this.offset = 0;
        this.waiting = this.PAUSE_FRAMES;
      }
    } else {
      // Right direction: scroll from max to 0
      this.offset -= this.SPEED;
      if (this.offset <= 0) {
        this.offset = this.scrollLimit;
        this.waiting = this.PAUSE_FRAMES;
      }
    }

    this.adjustment.set_value(this.offset);
  }
}

export const createMarquee = (
  className: string,
  direction: MarqueeDirection = 'left',
) => {
  return new MarqueeComponent(className, direction);
};

File: desktop_widgets/src/common/marquee.rs
use gtk4::prelude::*;
use gtk4::{Align, Box, Label, Orientation, ScrolledWindow, PolicyType, Adjustment};
use std::cell::RefCell;
use std::rc::Rc;
// use gtk4::glib; // Unused

#[derive(Clone)]
pub struct MarqueeLabel {
    pub container: ScrolledWindow,
    label1: Label,
    label2: Label,
    adjustment: Adjustment,
    #[allow(dead_code)] // Keep for potential direct access
    scroll_box: Box,
    state: Rc<RefCell<MarqueeState>>,
}

struct MarqueeState {
    text: String,
    offset: f64,
    waiting: i32,
    enabled: bool,
    scroll_limit: f64,
    direction: Direction,
}

#[derive(PartialEq, Clone, Copy)]
pub enum Direction {
    Left,
    Right,
}

impl MarqueeLabel {
    pub fn new(css_class: &str, direction: Direction, align: Align) -> Self {
        let label1 = Label::builder()
            .label("...")
            .halign(Align::Start)
            .wrap(false)
            .lines(1)
            .build();
        label1.add_css_class(css_class);

        let label2 = Label::builder()
            .label("...")
            .halign(Align::Start)
            .wrap(false)
            .lines(1)
            .visible(false) // Hidden by default
            .build();
        label2.add_css_class(css_class);

        let scroll_box = Box::builder()
            .orientation(Orientation::Horizontal)
            .spacing(60) // Gap between text repeats
            .halign(align) // Use dynamic alignment
            .build();
        
        scroll_box.append(&label1);
        scroll_box.append(&label2);

        let adjustment = Adjustment::new(0.0, 0.0, 10000.0, 1.0, 100.0, 0.0);

        let container = ScrolledWindow::builder()
            .hscrollbar_policy(PolicyType::External) // Hide scrollbar but allow scrolling
            .vscrollbar_policy(PolicyType::Never)
            .hadjustment(&adjustment)
            .propagate_natural_width(false) // CRITICAL: Prevent expansion
            .propagate_natural_height(false)
            .hexpand(true)
            .focusable(false)
            .can_focus(false)
            .can_target(false) // Let clicks pass through if possible
            .build();
        
        container.set_child(Some(&scroll_box));
        container.add_css_class("title-scroll");

        let state = Rc::new(RefCell::new(MarqueeState {
            text: String::new(),
            offset: 0.0,
            waiting: 100, // Frames to wait before scrolling
            enabled: false,
            scroll_limit: 0.0,
            direction,
        }));

        Self {
            container,
            label1,
            label2,
            adjustment,
            scroll_box,
            state,
        }
    }

    pub fn set_text(&self, text: &str) {
        let mut s = self.state.borrow_mut();
        if s.text == text {
            return;
        }
        s.text = text.to_string();
        
        self.label1.set_label(text);
        self.label2.set_label(text);
        
        self.reset(&mut s);
    }
    
    fn reset(&self, s: &mut MarqueeState) {
        s.offset = 0.0;
        s.waiting = 100;
        s.enabled = false;
        self.label2.set_visible(false);
        self.adjustment.set_value(0.0);
    }

    pub fn tick(&self) {
        let mut s = self.state.borrow_mut();
        
        // Init check if not enabled yet
        if !s.enabled {
            let container_width = self.container.allocation().width();
            if container_width > 0 {
                // Measure natural width of label 1
                let (_, nat_width, _, _) = self.label1.measure(Orientation::Horizontal, -1);
                
                if nat_width > container_width {
                     s.enabled = true;
                     s.scroll_limit = (nat_width + 60) as f64; // Width + Gap
                     self.label2.set_visible(true);
                     
                     if s.direction == Direction::Right {
                         s.offset = s.scroll_limit;
                         self.adjustment.set_value(s.offset);
                     }
                }
            }
            return;
        }

        if s.waiting > 0 {
            s.waiting -= 1;
            return;
        }

        let speed = 1.0; 

        if s.direction == Direction::Left {
            s.offset += speed;
            if s.offset >= s.scroll_limit {
                s.offset = 0.0;
                s.waiting = 100;
            }
        } else {
            // Right scroll (for Artist?)
            s.offset -= speed;
            if s.offset <= 0.0 {
                s.offset = s.scroll_limit;
                s.waiting = 100;
            }
        }

        self.adjustment.set_value(s.offset);
    }
}
